frontend/src/metabase/lib/expressions/recursive-parser.js

```diff
@@ -0,0 +1,274 @@
+import { tokenize, TOKEN, OPERATOR as OP } from "./tokenizer";
+
+import { getMBQLName, MBQL_CLAUSES } from "./index";
+
+const COMPARISON_OPS = [
+  OP.Equal,
+  OP.NotEqual,
+  OP.GreaterThan,
+  OP.LessThan,
+  OP.GreaterThanEqual,
+  OP.LessThanEqual,
+];
+
+function recursiveParse(source) {
+  const { tokens } = tokenize(source);
+
+  const peek = () => tokens[0];
+
+  const next = () => tokens.shift();
```

Comment:
Can you add a top level comment explaining how this parser moves through the token list? I was confused until `shift` caught my eye

Reply:
I've sprinkled some comments, including the BNF grammar whenever relevant.